File: src/server/happysad.pl
#! /usr/bin/env perl

##
## happysad.pl
##

use strict;
use warnings;
use feature qw(say);

use Getopt::Long;
use HappySadUnknown qw(glad_sad_or_unknown);

my ($input_text, $help);

GetOptions(
    'text|t=s' => \$input_text,
    'help'     => \$help,
);

help() if $help;
help("--text is required") if !$input_text;

say glad_sad_or_unknown($input_text);

exit 0;

## SUBS ##

sub help {
    say @_;
    say <<"    EOHELP";

    Name: 
        happysad.pl

    Usage:
        --text      required e.g. "I was glad to go to the beach. Very glad. 
                    Even though it was a miserable day."

        --help      this.

    EOHELP
    exit 0;
}

File: test/server/happysad.js
/*
 * test/server/happysad.js
 *
 */

import { expect } from 'chai';
import gladSadOrUnknown from '../../src/server/HappySadUnknown';

describe('HappySadUnknown.js', () => {
  describe('#gladSadOrUnknown', () => {
    var tests = [
      // edge cases
      { text: '', expected: 'unknown', name: 'empty string' },
      { text: '-', expected: 'unknown', name: 'punctuation only' },
      { text: undefined, expected: 'unknown', name: 'undefined' },

      // simple cases
      { text: 'glad', expected: 'happy', name: 'one happy word' },
      { text: 'miserable', expected: 'sad', name: 'one sad word' },
      { text: 'glad delighted sad miserable', expected: 'unknown', name: 'equal' },

      // less than 50% more
      {
        text: 'glad happy happy happy sad sad miserable',
        expected: 'unknown',
        name: 'more happy words but under 50% more'
      },
      {
        text: 'glad happy happy sad miserable sad miserable',
        expected: 'unknown',
        name: 'more sad words but under 50% more'
      },

      // over 50% more
      {
        text: 'glad happy happy happy sadmiserable',
        expected: 'happy',
        name: 'more happy words and over 50% more'
      },
      {
        text: 'glad happy sad miserable sad miserable',
        expected: 'sad',
        name: 'more sad words and over 50% more'
      },

      // exactly 50% more
      { 
        text: 'glad sad miserable',
        expected: 'sad',
        name: 'exactly 50% more sad than happy'
      },
      {
        text: 'I was glad to go to the beach. Very glad. Even though it was a miserable day.',
        expected: 'happy',
        name: 'provided test case'
      }
    ];
    for (var t of tests) {
      it(t.name, () => {
        expect(gladSadOrUnknown(t.text))
          .to.equal(t.expected);
      });
    }
  });
});
